src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObjectManaged.cs

```diff
@@ -36,15 +39,25 @@ protected override bool CheckPacket(PacketHandle packet, TaskCompletionSource<ob
         protected override void CreateSessionHandle(TdsParserStateObject physicalConnection, bool async)
         {
             Debug.Assert(physicalConnection is TdsParserStateObjectManaged, "Expected a stateObject of type " + GetType());
-            TdsParserStateObjectManaged managedSNIObject = physicalConnection as TdsParserStateObjectManaged;
-            _sessionHandle = managedSNIObject.CreateMarsSession(this, async);
-            SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.CreateSessionHandle | Info | State Object Id {0}, Session Id {1}", _objectID, _sessionHandle?.ConnectionId);
+            if (physicalConnection is TdsParserStateObjectManaged managedSNIObject)
+            {
+                _sessionHandle = managedSNIObject.CreateMarsSession(this, async);
+                SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.CreateSessionHandle | Info | State Object Id {0}, Session Id {1}", _objectID, _sessionHandle?.ConnectionId);
+            }
+            else
+            {
+                throw ADP.InvalidOperation("physicalConnection type is incorrect");
+            }
         }
 
         internal SNIMarsHandle CreateMarsSession(object callbackObject, bool async)
         {
             SqlClientEventSource.Log.TryTraceEvent("TdsParserStateObjectManaged.CreateMarsSession | Info | State Object Id {0}, Session Id {1}, Async = {2}", _objectID, _sessionHandle?.ConnectionId, async);
-            return _marsConnection.CreateMarsSession(callbackObject, async);
+            if (_marsConnection is null)
+            {
+                ThrowClosedConnection();
+            }
+            return _marsConnection!.CreateMarsSession(callbackObject, async);
```

Comment:
This non-null assertion suffix is required only for netstandard2.0 builds, on netcore 2.1 and later correctly understand the control flow and understand that the variable is not null.